### ISSUE TITLE
Investigate importance sampling for neural network training

### ISSUE THREAD

**Issue by JohannesGaessler** [ggml-org/ggml#1261]
Importance sampling is a Monte Carlo technique for variance reduction. When calculating an integral, the idea is to sample less frequently from regions that contribute little to the integral and to in return scale up the contributions from those regions. For neural network training, since the gradients are going to be proportional to the loss, the idea would be sample data with low loss less frequently in order to reduce the number of backward passes. Ideally this would still give you a good estimate of the gradients but with less computation.
